test/dash-playlist-loader.test.js

```diff
@@ -30,51 +30,149 @@ QUnit.test('throws if the playlist url is empty or undefined', function(assert)
   let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
 
-  assert.notOk(loader.started, 'not started');
   loader.load();
-  assert.equal(loader.state, 'HAVE_NOTHING', 'no metadata has loaded yet');
-  assert.ok(loader.started, 'started');
+  assert.equal(this.requests.length, 1, 'made a request');
+  assert.equal(this.requests[0].url, 'dash.mpd', 'requested the manifest');
+});
 
-  loader.master = { playlists: { 'playlist-1': { endList: true } }, mediaGroups: {} };
-  loader.parseMasterXml = () => {
-    return { playlists: [], mediaGroups: {} };
-  };
+QUnit.test('starts without any metadata', function(assert) {
+  let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
 
-  let newLoader =
-    new DashPlaylistLoader({ uri: 'playlist-1' }, this.fakeHls, false, loader);
+  assert.notOk(loader.started, 'not started');
 
-  assert.equal(newLoader.state, 'HAVE_METADATA', 'has metadata');
-  assert.ok(newLoader.started, 'started');
+  loader.load();
+  assert.equal(loader.state, 'HAVE_NOTHING', 'no metadata has loaded yet');
+  assert.ok(loader.started, 'started');
 });
 
-QUnit.test('requests the manifest immediately when given a URL', function(assert) {
+QUnit.skip('moves to HAVE_MASTER after loading a master playlist', function(assert) {
   let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
+  let state;
+
+  loader.on('loadedplaylist', function() {
+    state = loader.state;
+  });
 
   loader.load();
-  assert.equal(this.requests.length, 1, 'made a request');
-  assert.equal(this.requests[0].url, 'dash.mpd', 'requested the manifest');
+  standardXHRResponse(this.requests.shift());
+
+  assert.ok(loader.master, 'the master playlist is available');
+  assert.strictEqual(state, 'HAVE_MASTER', 'the state at loadedplaylist correct');
 });
 
-QUnit.test('moves to HAVE_MASTER and HAVE_METADATA after loading the manifest',
-function(assert) {
+QUnit.test('moves to HAVE_METADATA after loading a media playlist', function(assert) {
+  let loadedPlaylist = 0;
+  let loadedMetadata = 0;
   let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
-  let loadedPlaylistStates = [];
 
   loader.load();
 
   loader.on('loadedplaylist', function() {
-    loadedPlaylistStates.push(loader.state);
+    loadedPlaylist++;
+  });
+  loader.on('loadedmetadata', function() {
+    loadedMetadata++;
   });
+
   standardXHRResponse(this.requests.shift());
-  assert.ok(loader.master, 'the master playlist is available');
+  this.clock.tick(1);
+
   // because DASH only has one manifest, it should go through two loadedplaylists
   // and end with HAVE_METADATA because it already has the first media ready
-  assert.equal(loadedPlaylistStates.length, 2, 'triggered two loadedplaylist events');
-  assert.equal(loadedPlaylistStates[0], 'HAVE_MASTER', 'got master first');
-  assert.equal(loadedPlaylistStates[1], 'HAVE_METADATA', 'got media second');
+  assert.strictEqual(loadedPlaylist, 2,
```

Comment:
This is incorrect. The comment above is under the assumption that we are skipping the `HAVE_MASTER` state and jumping straight to `HAVE_METADATA`, but we discussed mapping the sidx to an hls child playlist conceptually. We should assert that (not using assert syntax cause I'm lazy)
```
loadedPlaylist === 1
loadedMetadata === 0
loaded.master !== undefined
loaded.media() === undefined
loader.state === HAVE_MASTER
```
this is the point at which initial playlist selection would be made, and then a call to `loader.media(initialSelectedPlaylist)` is made. After making that call, you would assert that
```
loadedPlaylist === 2
loadedMetadata === 1
loaded.media() !== undefined
loader.state === HAVE_METADATA
```
This series of assertions will be different than what is needed for the child loader below, which I'll touch on in a separate comment

Reply:
If there is no sidx though, wouldn't we want to make the initial playlist selection as part of `loader.load()`? The current asserts are expecting that the video playlist has been selected.

Reply:
PlaylistLoader also does this https://github.com/videojs/http-streaming/blob/master/src/playlist-loader.js#L543

Reply:
Only if `this.request` is not set. You can see in the line before, we trigger `loadedplaylist`, which has a listener here https://github.com/videojs/http-streaming/blob/master/src/master-playlist-controller.js#L222-L242 This will call `masterPlaylistLoader_.media()` which will result in a request for a media playlist be made, and thus `this.request` will be set and that pathway will not be taken.

So definitely ew want to make the initial selection for the `DashPlaylistLoader` in `load`, but only if `this.request` is *not* set. The big difference here is that for the `DashPlaylistLoader` request is now referring to either the sidx request or just an asynchronous `setTimeout` in the case we don't actually have a sidx but still wait to "fake" that we are "making a request".